src/Simulation/Simulators/SparseSimulator/Native/SparseSimulator.h

```diff
@@ -0,0 +1,851 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <string>
+#include <random>
+#include <cmath>
+#include <functional>
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <list>
+#include <set>
+
+#include "quantum_state.hpp"
+#include "basic_quantum_state.hpp"
+#include "types.h"
+#include "gates.h"
+
+using namespace std::literals::complex_literals;
+
+namespace Microsoft::Quantum::SPARSESIMULATOR
+{
+
+constexpr logical_qubit_id MAX_QUBITS = 1024;
+constexpr logical_qubit_id MIN_QUBITS = 64;
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
```

Comment:
Although I absolutely love the constexpr, I vote for #define in this case. M_PI has historically been #defined since "C" language times and there were widely varying rules to get to it. Only in C++20 there's a way to get to Pi constant via C++ methods. So until we use C++20, this is the most compatible way. See https://en.cppreference.com/w/cpp/header/numbers